sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,370 @@
+  // Please note that merging arbitrary plans can be complicated, the current version supports only
+  // some of the most important nodes.
+  private def tryMergePlans(
+      newPlan: LogicalPlan,
+      cachedPlan: LogicalPlan): Option[(LogicalPlan, AttributeMap[Attribute])] = {
+    checkIdenticalPlans(newPlan, cachedPlan).map(cachedPlan -> _).orElse(
+      (newPlan, cachedPlan) match {
+        case (np: Project, cp: Project) =>
+          tryMergePlans(np.child, cp.child).map { case (mergedChild, outputMap) =>
+            val (mergedProjectList, newOutputMap) =
+              mergeNamedExpressions(np.projectList, outputMap, cp.projectList)
+            val mergedPlan = Project(mergedProjectList, mergedChild)
+            mergedPlan -> newOutputMap
+          }
+        case (np, cp: Project) =>
+          tryMergePlans(np, cp.child).map { case (mergedChild, outputMap) =>
+            val (mergedProjectList, newOutputMap) =
+              mergeNamedExpressions(np.output, outputMap, cp.projectList)
+            val mergedPlan = Project(mergedProjectList, mergedChild)
+            mergedPlan -> newOutputMap
+          }
+        case (np: Project, cp) =>
+          tryMergePlans(np.child, cp).map { case (mergedChild, outputMap) =>
+            val (mergedProjectList, newOutputMap) =
+              mergeNamedExpressions(np.projectList, outputMap, cp.output)
+            val mergedPlan = Project(mergedProjectList, mergedChild)
+            mergedPlan -> newOutputMap
+          }
+        case (np: Aggregate, cp: Aggregate) if supportedAggregateMerge(np, cp) =>
+          tryMergePlans(np.child, cp.child).flatMap { case (mergedChild, outputMap) =>
+            val mappedNewGroupingExpression =
+              np.groupingExpressions.map(mapAttributes(_, outputMap))
+            // Order of grouping expression doesn't matter so we can compare sets
+            if (ExpressionSet(mappedNewGroupingExpression) ==
+              ExpressionSet(cp.groupingExpressions)) {
+              val (mergedAggregateExpressions, newOutputMap) =
+                mergeNamedExpressions(np.aggregateExpressions, outputMap, cp.aggregateExpressions)
+              val mergedPlan =
+                Aggregate(cp.groupingExpressions, mergedAggregateExpressions, mergedChild)
+              Some(mergedPlan -> newOutputMap)
+            } else {
+              None
+            }
+          }
+
+        // Merging general nodes is complicated. This implementation:
+        // - Supports only a whitelist of nodes (see `supportedMerge()`).
+        // - All children need to be direct fields of the tree node case class. Tree nodes where a
+        //   child is wrapped into a `Seq` or `Option` (e.g. `Union`) is not supported.
+        // - Expressions can be wrapped.
+        // - Children are tried to be merged in the same order.
+        case (np, cp) if supportedMerge(np) && np.getClass == cp.getClass &&
+            np.children.size == cp.children.size &&
```

Comment:
Well, I'm ok with dropping the general node merging code path for a more specific `Filter` and `Join` paths, but I think the current code can be easily extended with new nodes.

In this current PR `Filter`/`Join`s need to have the same condition. More precisely, what we test here in case of `Filter` and `Join` nodes is that:
- children of `newPlan` and `cachedPlan` nodes can be merged in order
- and mapped `newPlan` condition is the same as the condition of the `cachedPlan`
- and non-expression and non-child fields of the node case class should match (e.g. `Join.joinType`).

I added only the basic `Filter` and `Join` nodes to the whitelist for now, but I don't see any reason why some other nodes (`Sort`, `Window`, `Pivot`, ...) couldn't be added later.

Reply:
Fixed in https://github.com/apache/spark/pull/32298/commits/4da3fe64cf9822133e9651ec662a28dfdad246be.